File: benches/bench_serialize_and_deserialize.py
import numpy as np
import msgpack
import msgpack_numpy as m
import time

m.patch()


def bench_serialize(value, iterations):
    start_time = time.time()
    for _ in range(iterations):
        msgpack.packb(value)
    total_time = time.time() - start_time

    return total_time / iterations * 1000


def bench_deserialize(value, iterations):
    packed = msgpack.packb(value)

    start_time = time.time()
    for _ in range(iterations):
        msgpack.unpackb(packed)
    total_time = time.time() - start_time

    return total_time / iterations * 1000


if __name__ == "__main__":
    iterations = 10  # Adjust this number as needed

    # large arrays
    arr = np.arange(1000, dtype=np.float16)
    arrs = [arr for _ in range(10000)]

    # small but many arrays
    # arr = np.arange(100, dtype=np.float16)
    # arrs = [arr for _ in range(100000)]

    print(f"Running benchmarks with {iterations} iterations each...")

    serialize_avg = bench_serialize(arrs, iterations)
    print(f"Serialize average: {serialize_avg:.3f} ms")

    deserialize_avg = bench_deserialize(arrs, iterations)
    print(f"Deserialize average: {deserialize_avg:.3f} ms")

File: benches/bench_serialize_and_deserialize.rs
use half::f16;
use msgpack_numpy::NDArray;
use ndarray::Array1;
use serde::{Deserialize, Serialize};
use std::time::{Duration, Instant};

fn bench_serialize<T: Serialize>(value: &T, iterations: u32) -> Duration {
    let start = Instant::now();
    for _ in 0..iterations {
        rmp_serde::to_vec_named(&value).unwrap();
    }
    let total_duration = start.elapsed();
    total_duration / iterations
}

fn bench_deserialize<'a, T: Serialize + Deserialize<'a>>(value: &T, iterations: u32) -> Duration {
    let buf = rmp_serde::to_vec_named(&value).unwrap();

    let start = Instant::now();
    for _ in 0..iterations {
        let _deserialized: Vec<NDArray> = rmp_serde::from_slice(&buf).unwrap();
        // zero-copy deserialization if good alignment
        // use msgpack_numpy::CowNDArray;
        // let _deserialized: Vec<CowNDArray> = rmp_serde::from_slice(&buf).unwrap();
    }
    let total_duration = start.elapsed();
    total_duration / iterations
}

fn main() {
    let iterations = 10; // Adjust this number as needed

    // large arrays
    let arr = NDArray::F16(
        Array1::range(0., 1000., 1.)
            .into_dyn()
            .mapv(f16::from_f32)
            .into(),
    );
    let arrs = vec![arr; 10000];

    // small but many arrays
    // let arr = NDArray::F16(
    //    Array1::range(0., 1000., 1.)
    //         .into_dyn()
    //         .mapv(f16::from_f32)
    //         .into(),
    // );
    // let arrs = vec![arr; 100000];

    // test zero-copy deserialization
    // use ndarray::Array2;
    // let arr = NDArray::F16(
    //     Array2::zeros((1024, 2048))
    //         .into_dyn()
    //         .mapv(f16::from_f32)
    //         .into(),
    // );
    // let arrs = vec![arr; 10];

    println!("Running benchmarks with {} iterations each...", iterations);

    let serialize_avg = bench_serialize(&arrs, iterations);
    println!("Serialize average: {:?}", serialize_avg);

    let deserialize_avg = bench_deserialize(&arrs, iterations);
    println!("Deserialize average: {:?}", deserialize_avg);
}
